From c271516051f609c1cd96c734590034c2e29d3d36 Mon Sep 17 00:00:00 2001
From: Yan Markman <ymarkman@marvell.com>
Date: Thu, 22 Nov 2018 13:42:08 +0200
Subject: [PATCH 0802/1051] net: mvpp2: do not recycle buf with consumed
 headroom

IPsec/XFRM packet-buffer sent out over MVPP2 could be
shared or cloned or re-fragmented and therefore should not
be placed into recycle pool by TX-done.

In most of cases the packet indeed detected as non-recyclable.
But under traffic >0.5Gbps occasionally may come packet with all
conditions good for recycling.
In this case the buffer is put into recycle, then is get for RX
and causes for "Bad page state" panic.

This packet has skb-headroom =2.
Add the headroom checker into recycling decision (e.g. don't recycle
buffer if the headroom has been completely consumed).

Change-Id: Ie4dfc948f3a24ac02bb751ccbde519f62d1fc1cb
Signed-off-by: Yan Markman <ymarkman@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/61219
Tested-by: iSoC Platform CI <ykjenk@marvell.com>
Reviewed-by: Stefan Chulski <stefanc@marvell.com>
[Kevin: The original patch got from Marvell sdk10.0_19.06]
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 drivers/net/ethernet/marvell/mvpp2/mvpp2_main.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/drivers/net/ethernet/marvell/mvpp2/mvpp2_main.c b/drivers/net/ethernet/marvell/mvpp2/mvpp2_main.c
index 4f88734c5041..81fdccf16b8a 100644
--- a/drivers/net/ethernet/marvell/mvpp2/mvpp2_main.c
+++ b/drivers/net/ethernet/marvell/mvpp2/mvpp2_main.c
@@ -3266,6 +3266,9 @@ static int mvpp2_recycle_get_bm_id(struct sk_buff *skb)
 	/* Check if skb could be free */
 	if (skb_shared(skb) || skb_cloned(skb))
 		return -1;
+	/* WA: don't put to recycle the buffer with fully consumed headroom */
+	if (skb_headroom(skb) <= MVPP2_MH_SIZE)
+		return -1;
 	/* Get bm-pool-id */
 	hash &= MVPP2_RXTX_HASH_BMID_MASK;
 	if (hash >= MVPP2_BM_POOLS_NUM)
-- 
2.17.1

